test/model/message_test.dart

```diff
@@ -0,0 +1,425 @@
+  // For anything deeper than that, use `message_list_test.dart`.
+  late MessageListView messageList;
+  late int notifiedCount;
+
+  void checkNotified({required int count}) {
+    check(notifiedCount).equals(count);
+    notifiedCount = 0;
+  }
+  void checkNotNotified() => checkNotified(count: 0);
+  void checkNotifiedOnce() => checkNotified(count: 1);
+
+  /// Initialize [store] and the rest of the test state.
+  void prepare({Narrow narrow = const CombinedFeedNarrow()}) {
+    final stream = eg.stream();
+    subscription = eg.subscription(stream);
+    store = eg.store()
+      ..addStream(stream)..addSubscription(subscription);
+    connection = store.connection as FakeApiConnection;
+    notifiedCount = 0;
+    messageList = MessageListView.init(store: store, narrow: narrow)
+      ..addListener(() {
+        notifiedCount++;
+      });
+    check(messageList).fetched.isFalse();
+    checkNotNotified();
+  }
+
+  /// Perform the initial message fetch for [messageList].
+  ///
+  /// The test case must have already called [prepare] to initialize the state.
+  Future<void> prepareMessages(
+    List<Message> messages, {
+    bool foundOldest = false,
+  }) async {
+    connection.prepare(json:
+      newestResult(foundOldest: foundOldest, messages: messages).toJson());
+    await messageList.fetchInitial();
+    checkNotifiedOnce();
+  }
+
+  Future<void> addMessages(Iterable<Message> messages) async {
+    for (final m in messages) {
+      await store.handleEvent(MessageEvent(id: 0, message: m));
+    }
+    checkNotified(count: messageList.fetched ? messages.length : 0);
+  }
+
+  group('reconcileMessages', () {
+    test('from empty', () async {
+      prepare();
+      check(store.messages).isEmpty();
+      final messages = [
+        eg.streamMessage(),
+        eg.streamMessage(),
+        eg.dmMessage(from: eg.otherUser, to: [eg.selfUser]),
+      ];
+      store.reconcileMessages(messages);
+      check(store.messages).deepEquals({
+        for (final m in messages) m.id: m,
+      });
+    });
+
+    test('from not-empty', () async {
+      prepare();
+      final messages = [
+        eg.streamMessage(),
+        eg.streamMessage(),
+        eg.dmMessage(from: eg.otherUser, to: [eg.selfUser]),
+      ];
+      await addMessages(messages);
+      final newMessage = eg.streamMessage();
+      store.reconcileMessages([newMessage]);
+      check(store.messages).deepEquals({
+        for (final m in messages) m.id: m,
+        newMessage.id: newMessage,
+      });
+    });
+
+    test('new message does not clobber old on ID collision', () async {
+      prepare();
+      final message = eg.streamMessage(id: 1, content: '<p>foo</p>');
+      await addMessages([message]);
+      check(store.messages).deepEquals({1: message});
+      final newMessage = eg.streamMessage(id: 1, content: '<p>bar</p>');
+      store.reconcileMessages([newMessage]);
+      check(store.messages).deepEquals({1: message});
```

Comment:
Would a "shallow-equals `[message]`" check be accomplished by `.deepEquals([message])`? (As opposed to something like `jsonEquals`?)

Reply:
That would work because `Message` doesn't override `==`, but that feels a bit fragile or implicit.

Given there's just one expected, it could be written `.single.identical(message)` and that's probably clearest. A more general pattern would look like `.deepEquals([(it) => it.identical(message)])` (untested — but the idea is that `deepEquals` accepts condition callbacks as elements).